enterprise_access/apps/content_assignments/tasks.py

```diff
@@ -0,0 +1,74 @@
+"""
+Tasks for content_assignments app.
+"""
+
+import logging
+
+from celery import shared_task
+from django.apps import apps
+from requests.exceptions import HTTPError
+
+from enterprise_access.apps.api_client.lms_client import LmsApiClient
+from enterprise_access.tasks import LoggedTaskWithRetry
+
+from .constants import LearnerContentAssignmentStateChoices
+
+logger = logging.getLogger(__name__)
+
+
+class CreatePendingEnterpriseLearnerForAssignmentTaskBase(LoggedTaskWithRetry):  # pylint: disable=abstract-method
+    """
+    Base class for the create_pending_enterprise_learner_for_assignment task.
+
+    Provides a place to define retry failure handling logic.
+    """
+
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
+        """
+        If the task fails for any reaon (whether or not retries were involved), set the assignment state to errored.
+
+        Function signature documented at: https://docs.celeryq.dev/en/stable/userguide/tasks.html#on_failure
+        """
+        logger.error(f'"{task_id}" failed: "{exc}"')
+        learner_content_assignment_uuid = args[0]
+        learner_content_assignment_model = apps.get_model('content_assignments.LearnerContentAssignment')
+
+        try:
+            assignment = learner_content_assignment_model.objects.get(uuid=learner_content_assignment_uuid)
+            assignment.state = LearnerContentAssignmentStateChoices.ERRORED
+        except assignment.DoesNotExist:
+            logger.error(f'LearnerContentAssignment not found with UUID: {learner_content_assignment_uuid}')
+
+        if isinstance(exc, HTTPError):
+            # The failure resulted from too many retries.  This fact would be a useful thing to record in a "reason"
+            # field on the assignment if one existed.
+            logger.error('The task failure resulted from too many retries.')
```

Comment:
An LMS/enterprise API response with any 4xx/5xx status code will cause the celery task to retry.  If execution enters on_failure, then we can infer that it was the result of exceeding max_retries if the exc is any of the following:

https://github.com/openedx/enterprise-access/blob/c9681b7c18997630214ab94c29d79234a0770638/enterprise_access/tasks.py#L20-L26

If anything, I should probably write this like:

```suggestion
        if isinstance(exc, (RequestsConnectionError, RequestsTimeoutError, HTTPError, OperationalError)):
            # The failure resulted from too many retries.  This fact would be a useful thing to record in a "reason"
            # field on the assignment if one existed.
            logger.error('The task failure resulted from too many retries.')
```

So, as much as I'd like to refactor LoggedTaskWithRetry to be more specific about which specific status codes to treat as retryable, I think that's probably out of scope since it would either mean impacting all other consumers of LoggedTaskWithRetry, or adding configurability to LoggedTaskWithRetry that could fit into a different ticket/feature request.